src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+
+        ;; mark address in use over this range
+        (map-set pox-addr-reward-cycles
+            { pox-addr: pox-addr }
+            { first-reward-cycle: first-reward-cycle, num-cycles: num-cycles }
+        )
+
+        (ok true)
+    )
+)
+
+;; What is the minimum number of uSTX to be stacked in the given reward cycle?
+;; Used internally by the Stacks node, and visible publicly.
+(define-read-only (get-stacking-minimum (reward-cycle uint))
+    (let (
+        (ustx-stacked-so-far (get-total-ustx-stacked reward-cycle))
+    )
+    (begin
+        (if (< ustx-stacked-so-far (/ total-liquid-ustx u4))
+            ;; less than 25% of all liquid STX are stacked, so the threshold is smaller
+            (/ total-liquid-ustx STACKING-THRESHOLD-25)
+            ;; at least 25% of all liquid STX are stacked, so the threshold is larger
+            (/ total-liquid-ustx STACKING-THRESHOLD-100))
+    ))
+)
+
+;; Is the address mode valid for a PoX burn address?
+(define-private (check-pox-addr-version (version uint))
+    (and (>= version ADDRESS-VERSION-P2PKH)
+         (<= version ADDRESS-VERSION-P2WSH)))
+
+;; Is the given lock period valid?
+(define-private (check-pox-lock-period (lock-period uint)) 
+    (and (>= lock-period MIN-POX-REWARD-CYCLES) 
+         (<= lock-period MAX-POX-REWARD-CYCLES)))
+
+;; Register a PoX address for one or more reward cycles, and set how many uSTX it locks up initially.
+;; Will fail if the PoX address is already registered in one of the reward cycles.
+;; Will fail if the number of uSTX is beneath the lowest allowed Stacking threshold at the time of the call.
+(define-private (register-pox-addr-checked (pox-addr (tuple (version uint) (hashbytes (buff 20))))
+                                           (amount-ustx uint)
+                                           (first-reward-cycle uint)
+                                           (num-cycles uint)
+                                           (is-delegate bool))
+    (let (
+        (ustx-min (get-stacking-minimum first-reward-cycle))
+        (is-registered (is-pox-addr-registered pox-addr first-reward-cycle num-cycles))
+    )
+    (begin
+        ;; can't be registered yet if not a delegate
+        ;; (the delegate claims its PoX address when it registers)
+        (asserts! (or (not is-registered) is-delegate)
+            (err ERR-STACKING-POX-ADDRESS-IN-USE))
+
+        ;; minimum uSTX must be met
+        (asserts! (<= ustx-min amount-ustx)
+            (err ERR-STACKING-THRESHOLD-NOT-MET))
+
+        ;; lock period must be in acceptable range.
+        (asserts! (check-pox-lock-period num-cycles)
+            (err ERR-STACKING-INVALID-LOCK-PERIOD))
+
+        ;; address version must be valid
+        (asserts! (check-pox-addr-version (get version pox-addr))
+            (err ERR-STACKING-INVALID-POX-ADDRESS))
+
+        ;; register address and stacking
+        (add-pox-addr-to-reward-cycles pox-addr first-reward-cycle num-cycles amount-ustx)
```

Comment:
I had a chat with @kantai yesterday, many participants from the hackathon had similar flaws in their contracts.
The result here should be unwrap, otherwise you're shallowing errors returned by `add-pox-addr-to-reward-cycles ` and proceeding + returning true, instead of halting the execution of the contract.

Reply:
Thanks!  The only `(err ..)` that would be returned here is `ERR-STACKING-UNREACHABLE` (i.e. if this error is returned, the _whole chain_ should abort).  I'll make that happen here as well.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/2ad2a33bf91c71fffd2caa36b8708221708f5e15